graphql/query_resolver.go

```diff
@@ -982,44 +982,79 @@ func (r *queryResolver) Waterfall(ctx context.Context, options WaterfallOptions)
+	}
+
+	// Since GetAllWaterfallVersions uses an inclusive order range ($gte instead of $gt), add 1 to our minimum range
+	minVersionOrder := minOrderOpt + 1
+	if minOrderOpt == 0 {
+		minVersionOrder = activeVersions[len(activeVersions)-1].RevisionOrderNumber
+	}
+
+	// Same as above, but subtract for max order
+	maxVersionOrder := maxOrderOpt - 1
+	if maxOrderOpt == 0 {
+		maxVersionOrder = activeVersions[0].RevisionOrderNumber
 	}
 
-	versions, err := model.GetWaterfallVersions(ctx, projectId, opts)
+	allVersions, err := model.GetAllWaterfallVersions(ctx, projectId, minVersionOrder, maxVersionOrder)
 	if err != nil {
 		return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting waterfall versions: %s", err.Error()))
 	}
 
 	// TODO DEVPROD-10179: Add check to ensure each version has tasks that match filter...
 	// Something like this: https://github.com/evergreen-ci/evergreen/blob/bf8f12ec2eefe61f0cf9bcc594924c7be8f91d1b/graphql/query_resolver.go#L869-L938
-	// We should also be able to add a check to the version's build_variant_status field in case of a build variant filter (TODO DEVPROD-10178).
-	// For now, without filters it is guaranteed that `limit` matching versions have been returned.
+	// All other filters can be applied in the GetActiveWaterfallVersions pipeline, ensuring `limit` matching versions have been returned.
 
-	// TODO DEVPROD-10177: Implementing pagination will also allow us to fetch more versions if GetWaterfallVersions does not return enough *activated* versions
+	activeVersionIds := []string{}
+	for _, v := range activeVersions {
+		activeVersionIds = append(activeVersionIds, v.Id)
+	}
+
+	waterfallVersions := groupInactiveVersions(ctx, activeVersionIds, allVersions)
 
-	buildVariants, err := model.GetWaterfallBuildVariants(ctx, versions)
+	buildVariants, err := model.GetWaterfallBuildVariants(ctx, activeVersionIds)
 	if err != nil {
 		return nil, InternalServerError.Send(ctx, fmt.Sprintf("getting waterfall build variants: %s", err.Error()))
 	}
 
-	apiVersions := []*restModel.APIVersion{}
-	for _, v := range versions {
-		apiVersion := restModel.APIVersion{}
-		apiVersion.BuildFromService(v)
-		apiVersions = append(apiVersions, &apiVersion)
-	}
-
 	bv := []*model.WaterfallBuildVariant{}
 	for _, b := range buildVariants {
 		bCopy := b
 		bv = append(bv, &bCopy)
 	}
 
+	// If a max order was specified, use this as the new min and vice versa. If not, take this value from the bounds of the returned versions.
+	prevPageOrder := allVersions[0].RevisionOrderNumber
+	if maxOrderOpt > prevPageOrder {
+		prevPageOrder = maxOrderOpt
+	}
+	nextPageOrder := allVersions[len(allVersions)-1].RevisionOrderNumber
+	if minOrderOpt != 0 && minOrderOpt < nextPageOrder {
+		nextPageOrder = minOrderOpt
+	}
+	// If loading base page, there's no prev page to navigate to regardless of max order
+	if maxOrderOpt == 0 && minOrderOpt == 0 {
+		prevPageOrder = 0
+	}
+
 	return &Waterfall{
```

Comment:
would it be worth adding tests to `query_test.go` to test the resolver and verify that `nextPageOrder` and `prevPageOrder` are expected values (since otherwise I think they are untested)?